[BE-12] feat: JwtInterceptor 구현

xxbeann opened the pull request:
## #️⃣연관된 이슈

> ex) #33 

## 📝작업 내용

> 드디어 오랜 숙원이었던 JwtInterceptor를 구현했습니다.
총 3가지 메서드를 인터셉터에서 구현했습니다. 인터셉터를 구현하면서 JwtTokenProvider에 서비스 로직이 추가되었습니다.
1. CORS preflight 요청 -> 인터셉터 통과
2. 액세스토큰 재발급 요청에 대한 리프레시 토큰검증 -> 검증되면 인터셉터 통과
3. 로그아웃된 사용자의 요청 차단(사용자UUID를 통해 리프레시 토큰이 존재하는지 검증 후, 존재하지 않으면 에러 처리)

추가로 swagger 설정을 어노테이션 방식에서 bean 방식으로 리팩토링 했습니다.

### 스크린샷 (선택)

## 💬리뷰 요구사항(선택)

> 잘 봐주세요....

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * 로그아웃된 토큰 및 유효하지 않은 토큰 헤더에 대한 예외 처리 기능이 추가되었습니다.
  * JWT 토큰에서 토큰 추출 및 리프레시 토큰 검증 기능이 도입되었습니다.
  * JWT 토큰 유효성 검사 및 사용자 로그아웃 상태 확인을 위한 인터셉터가 도입되었습니다.
  * 사용자 ID 기반 리프레시 토큰 존재 여부 확인 및 예외 발생 기능이 추가되었습니다.

* **문서화**
  * Swagger(OpenAPI) 설정이 개선되어 보안 스키마 및 서버 정보가 동적으로 적용됩니다.

* **리팩터링**
  * 웹 요청에 대한 JWT 인터셉터가 WebConfig에 통합되어 인증 및 예외 처리가 일관되게 적용됩니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

xxbeann committed:
feat: JwtInterceptor 초기 구현

xxbeann committed:
feat: WebConfig에 JwtInterCeptor 추가

xxbeann committed:
Merge branch 'dev' into feat-#12

xxbeann committed:
Merge branch 'dev' into feat-#12

xxbeann committed:
feat: WebConfig에 Interceptor추가"

xxbeann committed:
feat: JWT Interceptor 초기 구현

xxbeann committed:
feat: SecuritySceme 어노테이션 방식으로 추가

xxbeann committed:
feat: swaggerInfo 어노테이션 방식에서 Bean 방식으로 변경

xxbeann committed:
feat: swagger 서버 루트 경로를 servletContex 기반으로 동적 설정

xxbeann committed:
feat: 보안 스키마 등록 및 API 전역 적용

xxbeann committed:
chore: 주석 삭제

xxbeann committed:
feat: CORS preflight 요청은 토큰 검증을 하지 않습니다.

xxbeann committed:
feat: extractToken 메서드 구현

Header를 떼고 body를 반환하는 extracToken 메서드를 구현했습니다.

xxbeann committed:
feat: validateRefreshToken 메서드 구현

인터셉터에서 사용되는 리프레시 토큰 유효성 검사 메서드입니다.
reissue api로 요청이 들어올 시, 리프레시 토큰이 유효한지 검사 후
인터셉터를 통과시켜 액세스 토큰을 재발급합니다.

xxbeann committed:
feat: 토큰 추출 후 refreshToken 유효성 검사

토큰 추출 후 refreshToken을 유효성 검사합니다.
이후 true 반환 시 인터셉터를 통과시킵니다.

xxbeann committed:
feat: userId 변수 정의

헤더로부터 추출한 token값을 통해 userId를 반환합니다.

xxbeann committed:
chore: 로그아웃 된 토큰 에러타입 정의

xxbeann committed:
feat: LoggedOutTokenException 커스텀 에러 정의

xxbeann committed:
feat: existByUserIdOrThrow 메서드 작성

로그아웃 된 사용자. 즉, 리프레시 토큰이 없는 사용자의 요청을 차단하기 위한 검증 함수입니다.

xxbeann committed:
refacotr: 접두어 제거 로직 삭제

xxbeann committed:
feat: JwtInterceptor 구현

xxbeann committed:
chore: 디버깅문 제거

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/JwtTokenProvider.java:
```diff
@@ -85,4 +85,23 @@ private Claims getClaimsFromToken(String token, TokenType tokenType) {
+    public String extractToken(String header) {
+        if (header == null || !header.startsWith(AUTH_TOKEN_HEADER)) {
+            throw new ExpiredTokenException();
+        }
```
헤더가 `null` 이거나 `Bearer`로 시작하지 않아서 던지는 커스텀 예외가 올바르지 않은 것 같습니다! `토큰 헤더가 없는 요청`이라는 예외를 던지는게 어떨까요?

xxbeann replied:
반영하겠습니다!!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/JwtTokenProvider.java:
```diff
@@ -85,4 +85,23 @@ private Claims getClaimsFromToken(String token, TokenType tokenType) {
+        if(!refreshTokenRepository.existsById(userId)) {
+            throw new LoggedOutTokenException();
+        }
+        return refreshTokenRepository.existsById(userId);
```
여기서는 굳이 Redis에 한번 더 접근하는 것보다 `true`를 반환하는게 어떨까요? 이미 앞에 있는 `if문`이 통과 됐다는 것은 `true`를 반환해도 되는 것이죠!

xxbeann replied:
반영하겠습니다!!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/common/config/JwtInterceptor.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+
+        // CORS Preflight 요청은 토큰 검증 X
+        if(CorsUtils.isPreFlightRequest(request)) {
+            return true;
+        }
+
+        String token = jwtTokenProvider.extractToken(request.getHeader(HttpHeaders.AUTHORIZATION));
+
+        // 액세스토큰 재발급 요청에 대한 리프레시 토큰검증
+        if (request.getRequestURI().startsWith(REISSUE_URI)) {
+            return jwtTokenProvider.validateRefreshToken(token);
+        }
+
+        Long userId = jwtTokenProvider.getUserIdFromAccessToken(token);
+
+        // 로그아웃된 사용자의 요청 차단
+        return jwtTokenProvider.existByUserIdOrThrow(userId);
```
`preHandle` 메소드가 현재 가지고 있는 책임이 preflight인지 확인, reissue 요청인지 확인, 로그아웃 사용자의 요청 차단으로 너무 많은 책임을 가지고 있습니다. 메소드 분리를 통해서 책임들을 분리시키는 게 어떨까요? 현재 preHandle에서는 각 책임들에 대한 로직이 어떻게 작동하는지 구체화 되어 있는데, 이 로직들을 각기 다른 메소드로 분리시키고 preHandle에서는 해당 메소드들을 호출하게끔 바꾸는 게 좋을 것 같습니다. 아래는 제가 생각한 예시입니다
``` java 

    if (isPreflightRequest(request)) {
        return true;
    }

    String token = extractToken(request);

    if (isReissueRequest(request)) {
        return validateRefreshToken(token);
    }

    if(isLoggedOutToken(token)) {
       return false;
    }
    
    return true;
}
```

xxbeann replied:
리팩토링 해보겠습니다~

Woojae-Jeong reviewed (commented):
전반적으로 잘 작성해주셨네요! 리뷰 몇 개 남겼는데 확인해주세요!

xxbeann committed:
Merge branch 'dev' into feat-#12

xxbeann committed:
Merge branch 'dev' into feat-#12

xxbeann committed:
feat: InvalidTokenHeader 커스텀 에러타입 정의

xxbeann committed:
feat: InvalidTokenHeader 커스텀 에러타입 정의

xxbeann committed:
refactor: 커스텀 에러 변경

xxbeann committed:
refacor: redis에 접근해 true값을 반환하는 것을 if문 통과 후 true 반환으로 변경

xxbeann committed:
Merge branch 'dev' into feat-#12

xxbeann committed:
refacor: preHandle 책임분리

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/JwtTokenProvider.java:
```diff
@@ -85,4 +86,23 @@ private Claims getClaimsFromToken(String token, TokenType tokenType) {
+        return true;
+    }
+
+    public boolean existByUserIdOrThrow(Long userId) {
+        if(!refreshTokenRepository.existsById(userId)) {
+            throw new LoggedOutTokenException();
+        }
+        return true;
```
해당 메소드의 기능은 사용자가 존재하는지 확인하는 역할입니다. 토큰을 발급하고 토큰에 담긴 정보를 추출하는 역할을 하는 `JwtTokenProvider`가 과연 `refreshToken`이 존재하는지 확인하는 책임을 가져야 하는지 고민해보면 좋을 것 같습니다.

Woojae-Jeong replied:
그리고 추가적으로 로그아웃에 해당되는 예외를 누가 던지는게 적절한지도 고민해보시면 좋을 것 같네요

xxbeann replied:
1. `existByUserIdOrThrow` 인증 상태를 검사하는 메서드라, authTokenService로 분리시키는게 더 나을까요 아니면 따로 클래스 분리를 시키는 게 좋을까요? 저는 전자가 더 나아보입니다!
2. 예외처리도 1번과 관련된 문제인 것 같은데, 다른 클래스로 분리시키거나 다른 서비스로 이동시키면 문제는 없어보입니다!

Woojae-Jeong replied:
1. 넵 저도 전자 방식이 올바르다고 생각해서 말씀드렸습니다!
2. 넵 맞습니다. 다만 혹시나 제 질문 의도를 잘못 이해할 수 있을 것 같아서 말씀을 좀 더 드리면, 예외를 던지는 쪽이 `Interceptor`일지 `JwtTokenProvider` 일지 `AuthTokenService` 일지 고민해보시면 좋을 것 같다고 말씀드렸습니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/common/config/JwtInterceptor.java:
```diff
@@ -0,0 +1,60 @@
+        return !isLoggedOutRequest(request);
+    }
+
+    public boolean isPreFlightRequest(HttpServletRequest request) {
+        return CorsUtils.isPreFlightRequest(request);
+    }
+
+    public boolean isReissueRequest(HttpServletRequest request) {
+        return request.getRequestURI().startsWith(REISSUE_URI);
+    }
+
+    public boolean validateRefreshToken(HttpServletRequest request) {
+        String token = extractToken(request);
+        return jwtTokenProvider.validateRefreshToken(token);
+    }
+
+    public boolean isLoggedOutRequest(HttpServletRequest request) {
+        Long userId = getUserIdFromAccessToken(request);
+        return jwtTokenProvider.existByUserIdOrThrow(userId);
+    }
+
+    public String extractToken(HttpServletRequest request) {
+        return jwtTokenProvider.extractToken(request.getHeader(HttpHeaders.AUTHORIZATION));
+    }
+
+    public Long getUserIdFromAccessToken(HttpServletRequest request) {
+        return jwtTokenProvider.getUserIdFromAccessToken(extractToken(request));
+    }
+}
```
현재 책임을 분리시킨 메소드의 접근제한자는 모두 `public`입니다. 어떤 메소드를 외부에 공개시키고, 어떤 메소드는 공개하지 않을 것인지 고민해보시면 좋을 것 같습니다

xxbeann replied:
내부에서만 쓰는 메서드는 감추는게 안정성에 도움이되는 군요??
리팩토링 하겠습니다!

Woojae-Jeong reviewed (commented):
코멘트 몇 개 남겼습니다! 한번 확인해보시고 답변 부탁드릴게요!

xxbeann committed:
refactor: 메서드 접근 제한자 변경

xxbeann committed:
refacor: existByUserIdOrThrow를 authTokenService로 책임 변경

xxbeann committed:
refactor:existByUserIdOrThrow 삭제

xxbeann committed:
feat: existByUserIdOrThrow 메서드 추가

xxbeann committed:
refactor: hasActiveRefreshToken 메서드 명 변경

xxbeann committed:
refactor: hasActiveRefreshToken 으로 메서드 명 변경

기존 existByUserIdOrThrow메서드명을 hasActiveRefreshToken으로 리팩토링 했습니다.
기존 메서드명보다 의도를 더 잘 담고있다고 생각해 변경하였습니다.

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/auth/jwt/service/AuthTokenService.java:
```diff
@@ -56,4 +57,13 @@ public void saveRefreshToken(Long userId, String refreshToken) {
     }
+
+    @Transactional(readOnly = true)
+    public boolean hasActiveRefreshToken(Long userId) {
+        boolean exists = refreshTokenRepository.existsById(userId);
+        if (!exists) {
+            throw new LoggedOutTokenException();
+        }
+        return true;
```
메소드이름을 `hasActiveRefreshToken`으로 지은 이유가 따로 있으실까요? 메소드명은 기능을 나타내는 역할을 하게 되는데, `hasActive` 라는 메소드명이 개인적으로 적절하지 않다고 생각을 합니다. 어찌보면 로그아웃된 토큰인지 검증하는 기능을 하는 것이라서 이를 잘 나타내는 메소드명으로 지어주는 게 좋아보입니다. 하지만 이는 개인적인 의견인 것이지 정답이 아니라서 좀 더 얘기를 듣고 싶어서 리뷰 남깁니다 

또한 결과값을 따로 `boolean` 타입에 저장하고, 결과값을 if문을 통해서 분기처리를 해도 되지만, 어차피 `refreshTokenRepository.existsById()` 메소드가 반환하는 것이 boolean타입이다보니 그대로 if문 안에 넣는 방식으로 코드를 설계하면 코드가 더 깔끔할 것 같네요!

xxbeann replied:
결국 로그아웃을 검증하려면 리프레시 토큰을 검증해야하기때문에 위와같이 메서드를 지었습니다. 또한, 나중에 다시 사용될 수도 있을 것 같아서 위와 같이 지었습니다..!
하지만 우재님 말씀대로 위 함수는 로그아웃이 되었는지를 검증하는 함수 입니다. `isLoggedIn` 정도로 메서드명 변경을 고려해 볼 수 있을 것 같습니다.
추가로 말씀해주신 분기처리는 리팩토링 해보겠습니다!

Woojae-Jeong replied:
좋습니다! 메소드명이 기능을 어찌보면 추상적으로 나타내는 역할이라서, 이를 잘 지어야지 다른 로직에서 사용할 때 헷갈리는 게 덜하더라구요! 이래서 개발자들이 변수명, 함수명 짓는 게 어렵다고 한다는 것 같네요 :)

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/common/config/JwtInterceptor.java:
```diff
@@ -32,29 +35,29 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
     }
 
-    public boolean isLoggedOutRequest(HttpServletRequest request) {
+    private boolean isLoggedOutRequest(HttpServletRequest request) {
         Long userId = getUserIdFromAccessToken(request);
-        return jwtTokenProvider.existByUserIdOrThrow(userId);
+        return authTokenService.hasActiveRefreshToken(userId);
     }
```
해당 메소드명과 내부 동작이 올바르지 않습니다. 해당 메소드는 로그아웃된 요청인지 체크하는 기능이며, 내부적으로 `hasActiveRefreshToken()`을 호출중입니다. 하지만 `hasActiveRefreshToken()` 은 로그아웃 된 토큰인 경우 예외를 던지고, 그렇지 않은 경우 true값을 리턴합니다. 그리고 그 결과값을 그대로 리턴하게 되는거죠. 즉, `isLoggedOutRequest` 메소드는 최종적으로 올바른 요청에 대해서 true 값을 리턴하게 됩니다. 메소드의 행위가 메소드명과 정반대로 이루어지고 있는 것이죠. 따라서 이에 대한 리팩토링이 필요해 보입니다

xxbeann replied:
이해했습니다! 
너무 헷갈리네요...
메서드를 `isLoggedInRequest`로 바꾸면 어떨까요? 어차피 인터셉터에서는 로그아웃된 요청을 막아야 하잖아요? 하지만 이는 `hasActiveRefreshToken`에서 True(로그인 상태)를 반환해버리기 때문에 메서드 명과 일치하지가 않더라구요...!
그래서, `isLoggedInRequest`로 리팩토링 후, 로그인 된 상태만 인터셉터를 통과하게 바꾸면 어떨까요?

Woojae-Jeong replied:
넵 저는 좋습니다! 어찌 되었든 간에 `로그아웃 된 요청은 거른다` 라는 핵심은 빗나가지 않으니까요!

Woojae-Jeong reviewed (commented):
코멘트 확인 부탁드립니다!

xxbeann committed:
refactor: isLoggedInRequest로 메서드명 변경

xxbeann committed:
refactor: isLoggedIn으로 메서드명 변경

xxbeann committed:
refactor: isLoggedIn 내부 로직 간소화

Woojae-Jeong commented:
확인했습니다! 머지시킬게요!
